### PR TITLE
MatrixFree diagonal: Avoid quadratic complexity & memory allocations

### PR DESCRIPTION
**Pull request by kronbichler**
This PR is inspired by looking at the MG global coarsening benchmark https://github.com/dealii/dealii/blob/master/tests/performance/timing_mg_glob_coarsen.cc, see also
mg_global_coarsening
(observe that I divide the setup time for the smoother by 2 to even fit on this plot). The main cause was that `ComputeDiagonalHelper` had quadratic complexity in the `submit` function: https://github.com/dealii/dealii/blob/master/include/deal.II/matrix_free/tools.h#L906-L921
For each diagonal entry we would add, we would search through all rows to see if they would give a contribution to the diagonal in some other position through the constraints. For unconstrained entries, we would only have a single entry, and even constrained faces in 3D only have a selection of DoFs. The right solution to this problem is to keep the inverse lookup between the column index (`i_comp`) and the affected rows in the present matrix column, which allows us to have a loop of length 1 (in the unconstrained case) or the underlying work in the constrained case, but not totally unnecessary integer work.

As a second improvement, this PR avoids frequent memory allocations by moving `std::vector` and `AlignedVector` fields out of the inner loops. We could do even better than I do here by making them members of `ComputeDiagonalHelper`, but already re-using them between different lanes of `VectorizedArray` helps significantly.

### CODE REVIEW

File: include/deal.II/matrix_free/tools.h [hunk 1]
```diff
@@ -836,8 +836,35 @@ namespace MatrixFreeTools
...
+
+              c_pool.inverse_lookup_rows.clear();
+              c_pool.inverse_lookup_rows.resize(1 + phi.dofs_per_cell);
+              for (const unsigned int i : c_pool.col)
+                c_pool.inverse_lookup_rows[1 + i]++;
+              // transform to offsets
+              for (unsigned int i = 1; i < c_pool.inverse_lookup_rows.size();
+                   ++i)
+                c_pool.inverse_lookup_rows[i] +=
+                  c_pool.inverse_lookup_rows[i - 1];
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/deal.II/matrix_free/tools.h:848]: Sounds good, I pushed that code.

**Review by peterrum — APPROVED** [review]

**Comment by kronbichler** [comment]
It might be of interest to see the performance test result of this PR. Using my own graphs for consistency with the first comment in this PR (instead of the "official" ones here: https://dealii.org/performance_tests/graphs/), this is what we see:
mg_global_coarsening
An improvement of a factor of 2, and there is a bit more to come with #15147 - after that, the computation of the diagonal will be dominated by the computation of the complete cell matrix (by applying the operator to all column vector), from which we throw everything away apart from the diagonal.

**Comment by bangerth** [comment]
Wow, very nice!

Do you happen what caused the deterioration in the computation of the rhs?

**Comment by kronbichler** [comment]
> Do you happen what caused the deterioration in the computation of the rhs?

The good thing about the way @tamiko set up the tests is that we run time for each merge commit. This shows:
```
# Date:      2023-04-18 19:30:27+00:00
# Site:      whistler-node
# Branch:    master
# Revision:  55ce8745a62148878b9b99f875b6acabb8ce46cb
# Timestamp: 2023-04-18 20:54:56+02:00
```
and this commit 55ce874 is from #15035. Not totally unexpected because it changes the evaluation of points in `MappingQ` used inside `VectorTools::interpolate`, which computes quite some information like the Jacobian: https://github.com/dealii/dealii/blob/d4d25625170f732fbd1f0640548e003d75ff1114/include/deal.II/numerics/vector_tools_interpolate.templates.h#L312-L321
On my machine, this PR actually lead to slight improvement, 0.42s -> 0.40s (128 Core AMD Epyc 7742 Rome), and a later PR #15125 improved this slightly more to 0.37s, but I do not run the tests regularly so I can't say for sure. Either way, I think this is something I've been looking at and probably not critical. (If we wanted to improve the benchmark, we would set more conservative mapping flags).

**Comment by bangerth** [comment]
OK, thanks for checking!

**Pull request merged**